QueryRow().Scan() fails for query with binding variables

<!--
SPDX-FileCopyrightText: 2020 SAP SE
SPDX-FileCopyrightText: 2021 SAP SE

SPDX-License-Identifier: Apache-2.0
-->

**Description**

QueryRow().Scan() fails for query with binding variables

**Expected behaviour**

QueryRow().Scan() should return correct result

**Current behaviour**

It returns error:
```
go-ase: cursor deletion finished with error: tds: error in user-defined processing function: query failed with errors: received EED messages: 557: The cursor '(Unknown name)' can not be used as it could not be found. It is possible that either it was not declared or it is not available in the current context.;
```

**Steps to Reproduce**

Steps to reproduce the behavior:
1. Export the following environment variables:
export DB_USERNAME=
export DB_PASSWORD=
export DB_DATABASE=
export DB_HOSTNAME=
export DB_PORT=
export DB_DRIVER=ase
2. Prepare test data
```
create table TEST_PROFILE(UserId BIGINT, ItemName varchar(32) not null, ItemValue varchar(32) not null)
create unique clustered index iuc_test_profile on TEST_PROFILE(UserId, ItemName)
insert into TEST_PROFILE values(1, "CUR_ACCOUNT", "1200")
```
3. Run 
```
DB_DRIVER=ase go test  -v
```

4. See error
```
=== RUN   TestQueryRowWithoutBindingVars
--- PASS: TestQueryRowWithoutBindingVars (0.10s)
=== RUN   TestQueryRowWithBindingVars
    queryrow_test.go:41: querying failed: go-ase: cursor deletion finished with error: tds: error in user-defined processing function: query failed with errors: received EED messages: 557: The cursor '(Unknown name)' can not be used as it could not be found. It is possible that either it was not declared or it is not available in the current context.;
--- FAIL: TestQueryRowWithBindingVars (0.10s)
FAIL
```
5. See the correct results from tds driver for comparison:
```
DB_DRIVER=tds go test -v
=== RUN   TestQueryRowWithoutBindingVars
--- PASS: TestQueryRowWithoutBindingVars (0.03s)
=== RUN   TestQueryRowWithBindingVars
--- PASS: TestQueryRowWithBindingVars (0.03s)
PASS
ok  	main	0.075s
```

**Environment**

Client: 
  OS: Centos-7
  go version: 1.16.3
  go-ase version / commit hash: v0.0.0-20210401085605-ce408ad54d8b
  go-dblib version / commit hash: v0.0.0-20210315102028-46975ce12d39

Server:
  OS: Centos-7
  ASE open client version: 16.0_SP03PL02_EBF28306.el7
  ASE data server version: 16.0_SP03PL09_EBF29247.r7.x86_64


**Additional context**


[queryrow_test.go.gz](https://github.com/SAP/go-ase/files/6388073/queryrow_test.go.gz)

--- a/cursor.go
+++ b/cursor.go
@@ -88,7 +88,12 @@ func (cursor *Cursor) allocateOnServer(ctx context.Context, query string, args [
 		if err != nil {
 			return fmt.Errorf("error creating stmt: %w", err)
 		}
+		// The cursor needs a reference to the statement to close it.
+		// The statement needs a reference to the cursor to mark it as
+		// already closed if the server decides to close the cursor
+		// immediately instead of waiting on being instructed to do so.
 		cursor.stmt = stmt
+		stmt.cursor = cursor
 	}
 
 	// If a cursor has arguments a statement with the query must be
@@ -292,6 +297,12 @@ func (cursor *Cursor) Close(ctx context.Context) error {
 		// cursor has an associated prepared statement
 		if err := cursor.stmt.Close(); err != nil {
 			return fmt.Errorf("go-ase: error closing stmt: %w", err)
+		}
+
+		// Return early if the cursor itself was automatically closed
+		// after deallocating the associated statement.
+		if cursor.closed {
+			return nil
 		}
 	}
 

--- a/dynamic.go
+++ b/dynamic.go
@@ -36,6 +36,8 @@ type Stmt struct {
 
 	paramFmt *tds.ParamFmtPackage
 	rowFmt   *tds.RowFmtPackage
+
+	cursor *Cursor
 }
 
 // Prepare implements the driver.Conn interface.
@@ -156,8 +158,30 @@ func (stmt *Stmt) close(ctx context.Context) error {
 		return err
 	}
 
-	if err := stmt.recvDoneFinal(ctx); err != nil {
-		return err
+	_, err := stmt.conn.Channel.NextPackageUntil(ctx, true, func(pkg tds.Package) (bool, error) {
+		switch typed := pkg.(type) {
+		case *tds.CurInfoPackage:
+			if typed.Command != tds.TDS_CUR_CMD_INFORM {
+				return true, fmt.Errorf("received %T with command %s instead of TDS_CUR_CMD_INFORM",
+					typed, typed.Command)
+			}
+
+			if typed.Status&tds.TDS_CUR_ISTAT_CLOSED == tds.TDS_CUR_ISTAT_CLOSED {
+				stmt.cursor.closed = true
+			}
+			return false, nil
+		case *tds.DonePackage:
+			if typed.Status != tds.TDS_DONE_FINAL {
+				return false, fmt.Errorf("DonePackage does not have status TDS_DONE_FINAL set: %s", typed)
+			}
+
+			return true, io.EOF
+		default:
+			return true, fmt.Errorf("unhandled package type %T: %s", typed, typed)
+		}
+	})
+	if err != nil && !errors.Is(err, io.EOF) {
+		return fmt.Errorf("error handling response to stmt deallocation: %w", err)
 	}
 
 	return nil

--- a/dynamicRecv.go
+++ b/dynamicRecv.go
@@ -36,26 +36,3 @@ func (stmt Stmt) recvDynAck(ctx context.Context) error {
 
 	return nil
 }
-
-func (stmt Stmt) recvDoneFinal(ctx context.Context) error {
-	_, err := stmt.conn.Channel.NextPackageUntil(ctx, true,
-		func(pkg tds.Package) (bool, error) {
-			done, ok := pkg.(*tds.DonePackage)
-			if !ok {
-				return false, nil
-			}
-
-			if done.Status != tds.TDS_DONE_FINAL {
-				return false, fmt.Errorf("DonePackage does not have status TDS_DONE_FINAL set: %s", done)
-			}
-
-			return true, io.EOF
-		},
-	)
-
-	if err != nil && !errors.Is(err, io.EOF) {
-		return err
-	}
-
-	return nil
-}

--- a/integration_cursorClose_test.go
+++ b/integration_cursorClose_test.go
@@ -1,0 +1,274 @@
+// SPDX-FileCopyrightText: 2021 SAP SE
+//
+// SPDX-License-Identifier: Apache-2.0
+
+// +build integration
+
+package ase
+
+import (
+	"context"
+	"database/sql"
+	"database/sql/driver"
+	"errors"
+	"fmt"
+	"io"
+	"testing"
+
+	"github.com/SAP/go-dblib/integration"
+)
+
+func TestCursorClose(t *testing.T) {
+
+	t.Run("SingeCursor", func(t *testing.T) {
+		integration.TestForEachDB("TestCursorCloseSingleCursor", t, func(t *testing.T, db *sql.DB, tableName string) {
+			wrapper(t, db, tableName, singleCursor)
+		})
+	})
+
+	t.Run("SingeCursorWithArgs", func(t *testing.T) {
+		integration.TestForEachDB("TestCursorCloseSingleCursorWithArgs", t, func(t *testing.T, db *sql.DB, tableName string) {
+			wrapper(t, db, tableName, singleCursorWithArgs)
+		})
+	})
+
+	t.Run("TwoCursors", func(t *testing.T) {
+		integration.TestForEachDB("TestCursorCloseTwoCursors", t, func(t *testing.T, db *sql.DB, tableName string) {
+			wrapper(t, db, tableName, twoCursors)
+		})
+	})
+
+	t.Run("TwoCursorsOneWithArgs", func(t *testing.T) {
+		integration.TestForEachDB("TestCursorCloseTwoCursorsOneWithArgs", t, func(t *testing.T, db *sql.DB, tableName string) {
+			wrapper(t, db, tableName, twoCursorsOneWithArgs)
+		})
+	})
+
+	t.Run("TwoCursorsWithArgs", func(t *testing.T) {
+		integration.TestForEachDB("TestCursorCloseTwoCursorsWithArgs", t, func(t *testing.T, db *sql.DB, tableName string) {
+			wrapper(t, db, tableName, twoCursorsWithArgs)
+		})
+	})
+
+	t.Run("TwoCursorsOneWithArgs2", func(t *testing.T) {
+		integration.TestForEachDB("TestCursorCloseTwoCursorsOneWithArgs2", t, func(t *testing.T, db *sql.DB, tableName string) {
+			wrapper(t, db, tableName, twoCursorsOneWithArgs2)
+		})
+	})
+
+}
+
+func createTable(db *sql.DB, tableName string) error {
+	if _, err := db.Exec("create table " + tableName + " (a int, b varchar(30))"); err != nil {
+		return fmt.Errorf("error creating table %s: %w", tableName, err)
+	}
+
+	for i, val := range []string{"one", "two", "three", "four"} {
+		if _, err := db.Exec("insert into "+tableName+" (a, b) values (?, ?)", i+1, val); err != nil {
+			return fmt.Errorf("error inserting values (%d, %s) into table %s: %w", i+1, val, tableName, err)
+		}
+	}
+
+	return nil
+}
+
+func wrapper(t *testing.T, db *sql.DB, tableName string, runner func(*testing.T, *Conn, string)) {
+	if err := createTable(db, tableName); err != nil {
+		t.Errorf("error creating table: %v", err)
+		return
+	}
+
+	conn, err := db.Conn(context.Background())
+	if err != nil {
+		t.Errorf("error getting conn from sql.DB: %v", err)
+		return
+	}
+	defer func() {
+		if err := conn.Close(); err != nil {
+			t.Errorf("error closing conn from sql.DB: %v", err)
+		}
+	}()
+
+	conn.Raw(func(driverConn interface{}) error {
+		aseConn, ok := driverConn.(*Conn)
+		if !ok {
+			t.Errorf("received driverConn is not *Conn: %v", err)
+			return nil
+		}
+
+		runner(t, aseConn, tableName)
+		return nil
+	})
+}
+
+func fetch(t *testing.T, cursor *Cursor) {
+	rows, err := cursor.Fetch(context.Background())
+	if err != nil {
+		t.Errorf("error fetching result set: %v", err)
+		return
+	}
+	defer func() {
+		if err := rows.Close(); err != nil {
+			t.Errorf("error closing rows: %v", err)
+		}
+	}()
+
+	values := []driver.Value{0, ""}
+	for {
+		if err := rows.Next(values); err != nil {
+			if errors.Is(err, io.EOF) {
+				break
+			}
+			t.Errorf("error reading row: %v", err)
+			return
+		}
+
+		fmt.Printf("| %d | %s |\n", values[0], values[1])
+	}
+
+	if err := rows.Close(); err != nil {
+		t.Errorf("error closing rows: %v", err)
+	}
+}
+
+func singleCursor(t *testing.T, conn *Conn, tableName string) {
+	cursor, err := conn.NewCursor(context.Background(), "select * from "+tableName)
+	if err != nil {
+		t.Errorf("error creating cursor: %v", err)
+		return
+	}
+	defer func() {
+		if err := cursor.Close(context.Background()); err != nil {
+			t.Errorf("error closing cursor: %v", err)
+		}
+	}()
+
+	fetch(t, cursor)
+}
+
+func singleCursorWithArgs(t *testing.T, conn *Conn, tableName string) {
+	cursor, err := conn.NewCursor(context.Background(), "select * from "+tableName+" where b like (?)", "two")
+	if err != nil {
+		t.Errorf("error creating cursor: %v", err)
+		return
+	}
+	defer func() {
+		if err := cursor.Close(context.Background()); err != nil {
+			t.Errorf("error closing cursor: %v", err)
+		}
+	}()
+
+	fetch(t, cursor)
+}
+
+func twoCursors(t *testing.T, conn *Conn, tableName string) {
+	cursor, err := conn.NewCursor(context.Background(), "select * from "+tableName)
+	if err != nil {
+		t.Errorf("error creating cursor: %v", err)
+		return
+	}
+	defer func() {
+		if err := cursor.Close(context.Background()); err != nil {
+			t.Errorf("error closing cursor: %v", err)
+		}
+	}()
+
+	fetch(t, cursor)
+
+	cursor2, err := conn.NewCursor(context.Background(), "select * from "+tableName)
+	if err != nil {
+		t.Errorf("error creating cursor2: %v", err)
+		return
+	}
+	defer func() {
+		if err := cursor2.Close(context.Background()); err != nil {
+			t.Errorf("error closing cursor2: %v", err)
+		}
+	}()
+
+	fetch(t, cursor2)
+}
+
+func twoCursorsOneWithArgs(t *testing.T, conn *Conn, tableName string) {
+	cursor, err := conn.NewCursor(context.Background(), "select * from "+tableName)
+	if err != nil {
+		t.Errorf("error creating cursor: %v", err)
+		return
+	}
+	defer func() {
+		if err := cursor.Close(context.Background()); err != nil {
+			t.Errorf("error closing cursor: %v", err)
+		}
+	}()
+
+	fetch(t, cursor)
+
+	cursorWithArgs, err := conn.NewCursor(context.Background(), "select * from "+tableName+" where b like (?)", "two")
+	if err != nil {
+		t.Errorf("error creating cursorWithArgs: %v", err)
+		return
+	}
+	defer func() {
+		if err := cursorWithArgs.Close(context.Background()); err != nil {
+			t.Errorf("error closing cursorWithArgs: %v", err)
+		}
+	}()
+
+	fetch(t, cursorWithArgs)
+}
+
+func twoCursorsOneWithArgs2(t *testing.T, conn *Conn, tableName string) {
+	cursorWithArgs, err := conn.NewCursor(context.Background(), "select * from "+tableName+" where b like (?)", "two")
+	if err != nil {
+		t.Errorf("error creating cursorWithArgs: %v", err)
+		return
+	}
+	defer func() {
+		if err := cursorWithArgs.Close(context.Background()); err != nil {
+			t.Errorf("error closing cursorWithArgs: %v", err)
+		}
+	}()
+
+	fetch(t, cursorWithArgs)
+
+	cursor, err := conn.NewCursor(context.Background(), "select * from "+tableName)
+	if err != nil {
+		t.Errorf("error creating cursor: %v", err)
+		return
+	}
+	defer func() {
+		if err := cursor.Close(context.Background()); err != nil {
+			t.Errorf("error closing cursor: %v", err)
+		}
+	}()
+
+	fetch(t, cursor)
+}
+
+func twoCursorsWithArgs(t *testing.T, conn *Conn, tableName string) {
+	cursorWithArgs, err := conn.NewCursor(context.Background(), "select * from "+tableName+" where b like (?)", "two")
+	if err != nil {
+		t.Errorf("error creating cursorWithArgs: %v", err)
+		return
+	}
+	defer func() {
+		if err := cursorWithArgs.Close(context.Background()); err != nil {
+			t.Errorf("error closing cursorWithArgs: %v", err)
+		}
+	}()
+
+	fetch(t, cursorWithArgs)
+
+	cursorWithArgs2, err := conn.NewCursor(context.Background(), "select * from "+tableName+" where b like (?)", "two")
+	if err != nil {
+		t.Errorf("error creating cursorWithArgs2: %v", err)
+		return
+	}
+	defer func() {
+		if err := cursorWithArgs2.Close(context.Background()); err != nil {
+			t.Errorf("error closing cursorWithArgs2: %v", err)
+		}
+	}()
+
+	fetch(t, cursorWithArgs2)
+}

